shub/image/list.py

```diff
@@ -89,27 +102,47 @@ def _get_project_settings(project, endpoint, apikey):
     return {k: v for k, v in req.json().items() if k in SETTING_TYPES}
 
 
-def _run_list_cmd(project, image_name, project_settings):
-    """Run `scrapy list` command inside the image container."""
-
+def _run_cmd_in_docker_container(image_name, command, environment):
+    """Run a command inside the image container."""
     client = utils.get_docker_client()
-    # FIXME we should pass some value for SCRAPY_PROJECT_ID anyway
-    # to handle `scrapy list` cmd properly via sh_scrapy entrypoint
-    project = str(project) if project else ''
-    job_settings = json.dumps(project_settings)
     container = client.create_container(
         image=image_name,
-        command=['list-spiders'],
-        environment={'SCRAPY_PROJECT_ID': project,
-                     'JOB_SETTINGS': job_settings})
+        command=[command],
+        environment=environment,
+    )
     if 'Id' not in container:
-        raise shub_exceptions.ShubException(
-            "Create container error:\n %s" % container)
-
-    client.start(container)
+        raise ShubException("Create container error:\n %s" % container)
+    try:
+        client.start(container)
+    except docker.errors.NotFound:
+        # return a proper exit code if executable is not found
+        return 127, None
     statuscode = client.wait(container=container['Id'])
-
     return statuscode, client.logs(
             container=container['Id'],
             stdout=True, stderr=True if statuscode else False,
             stream=False, timestamps=False)
+
+
+def _extract_metadata_from_image_info_output(output):
+    """Extract and validate spiders list from `shub-image-info` output."""
+
+    def raise_shub_image_info_error(error):
+        """Helper to raise ShubException with prefix and output"""
+        msg = "shub-image-info: {} \n[output '{}']".format(error, output)
+        raise ShubException(msg)
+
+    try:
+        metadata = json.loads(output)
+        spiders_list = metadata.get('spiders', [])
+    except (ValueError, AttributeError):
+        raise_shub_image_info_error('output is not a valid JSON dict')
+    if not isinstance(spiders_list, list):
+        raise_shub_image_info_error('spiders section must be a list')
+
+    spiders, scripts = [], []
+    for name in spiders_list:
+        if not (name and isinstance(name, string_types)):
+            raise_shub_image_info_error("spider name can't be empty or non-string")
+        scripts.append(name[3:]) if name.startswith('py:') else spiders.append(name)
+    return {'spiders': spiders, 'scripts': scripts}
```

Comment:
For `list-spiders` (and in the fallback to it in line 84 above), we ran the spider list through `utils.valid_spiders()`. Is there a reason to skip it here?